Add material about containers to CI/CD lesson (#18)
SS-NES/SS-NES-lesson: episodes/continuous-integration.md

CICD relies on version control.
Workflows are usually triggered by events, such as merging to the main branch happening on the verson control repository.

> # Docker Containers

Reviewer: ```suggestion
# Containers
```
Maybe call this section just Containers, as we mention also other alternatives to docker.

Nowadays software is often distributed or deployed using containers.
These containers contain every dependency an application needs and can run anywhere.
This is especially useful in applications that run as a service on a cloud provider such as Google or Amazon Web Services where it is not known before hand where an application will run.
These containers are light weight operating system images in which the application is stored including everything it needs.
To build these container images, CI/CD is extremely useful as it can be automated as explained below in the section explaining pipelines and workflows.
Running applications in containers tends to enforce decoupling from external dependencies and communicate to external services through well defined and stable interfaces.
Building and distributing containers is generally done using [docker](https://www.docker.com) but there are others such as [podman](https://podman.io/)

> 

Reviewer: ```suggestion
You can find more information about containers [here](https://book.the-turing-way.org/reproducible-research/renv/renv-containers.html).
```
Turing Way has a fairly detailed chapter on containers (using docker, podman and singularity). Maybe we can point to it as further reading.

Click on that to download the package.

That's it! Package published!

## Considerations

CI/CD pipelines are not very suitable if your tests requires a lot of static data.
Running large integration tests inside a CI/CD pipeline is thus not recommended as there is generally limited space and time in CI/CD pipelines.

> Writing small and fast unit tests that run automatically inside CI/CD pipelines rather than large integration tests is recommended.

Reviewer 2: I see the point of limiting CI/CD usage depending on available resources, but one of the benefits I see of CI/CD is exactly for running longer integration tests automatically: If I have some longer integration tests that might take a few minutes, I don't want to locally run my test suite each time I make a commit. Having those still run by CI/CD makes sure they're still used, even if I don't frequently/ever run them locally.

I would probably set it up such that it might only run e.g. before a release though, which I can see is a more advanced use case. Do you think there is space to be more (explicitly) nuanced about this? Or would it only cause confusion?

Author: Ik heb wat nuance aangebrach in 5105cc6.

Reviewer: For users who run into this type of issues, maybe it is worth pointing that there are other alternatives, like self-hosted runners.

Here are some links which we could point people to:
https://github.com/ci-for-research/self-hosted-runners https://tilburgsciencehub.com/topics/automation/automation-tools/deployment/ghactions-self-hosted-runner/

I imagine 90% of people who are just getting started won't need this, but maybe good for them to know that if they ever run into issues of tests taking too long, there are alternatives.

Author: Good point. Added in commit 5880255

> It is therefore helpful to practice sotware engineering best practices such as decoupling.
> Decoupling enables testable code.

Reviewer 2: ```suggestion
It is therefore helpful to practice software engineering best practices such as decoupling, since that will lead to more easily testable code.
```
